File: Template/Backend/Source/Base/Dto/AuditableDto.cs
namespace Backend.Base.Dto;

public abstract class AuditableDto<TId> : AuditableDto, IBaseDto<TId>
{
    public TId Id { get; set; } = default!;
}

public abstract class AuditableDto : IAuditableDto
{
    public DateTime CreatedAt { get; set; }
    public Guid? CreatedBy { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public Guid? UpdatedBy { get; set; }
}

public abstract class CreatableDto<TId> : CreatableDto, IBaseDto<TId>
{
    public TId Id { get; set; } = default!;
}

public abstract class CreatableDto : ICreatableDto
{
    public DateTime CreatedAt { get; set; }
    public Guid? CreatedBy { get; set; }
}

public abstract class UpdatableDto<TId> : UpdatableDto, IBaseDto<TId>
{
    public TId Id { get; set; } = default!;
}

public abstract class UpdatableDto : IUpdatableDto
{
    public DateTime? UpdatedAt { get; set; }
    public Guid? UpdatedBy { get; set; }
}

File: src/frontend/web/store/api/base/dto/auditable-dto.ts
import { BaseDto } from './base-dto'

/** Common creation audit fields (timestamp and creator) shared by auditable DTOs. */
export interface CreatableDto {
  createdAt: string
  createdBy: string
}

/** Common update audit fields (timestamp and last updater) shared by auditable DTOs. */
export interface UpdatableDto {
  updatedAt: string
  updatedBy: string
}

/** DTO that captures both creation and last-update audit information. */
export interface AuditableDto extends CreatableDto, UpdatableDto { }

/** Identifier-typed DTO that also includes creation audit fields. */
export interface GenericCreatableDto<TId> extends CreatableDto, BaseDto<TId> { }

/** Identifier-typed DTO that also includes last-update audit fields. */
export interface GenericUpdatableDto<TId> extends UpdatableDto, BaseDto<TId> { }

/** Identifier-typed DTO that includes both creation and last-update audit fields. */
export interface GenericAuditableDto<TId> extends AuditableDto, BaseDto<TId> { }
